File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int N = nums.length;
        int left = 0, right = 0;
        int zeros = 0, maxLen = 0;

        while(right < N){
            if(nums[right] == 0){
                zeros++;
            }
            while(zeros > k){
                if(nums[left] == 0){
                    zeros--;
                }
                left++;
            }
            int length = right - left + 1;
            maxLen = Math.max(maxLen, length);
            right ++;
        }
        return maxLen;
    }
}

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        N = len(nums)
        left, right = 0, 0
        zeros, maxLen = 0, 0

        while right < N:
            if nums[right] == 0:
                zeros += 1
            if zeros > k:
                if nums[left] == 0:
                    zeros -= 1
                left += 1
            if zeros <= k:
                length = right - left + 1
                maxLen = max(maxLen, length)
            right += 1
        return maxLen